error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/MonoZipOfMonoVoidUsage.java

```diff
@@ -0,0 +1,144 @@
+import static tech.picnic.errorprone.utils.MoreTypes.generic;
+import static tech.picnic.errorprone.utils.MoreTypes.type;
+
+import com.google.auto.service.AutoService;
+import com.google.common.collect.Iterables;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.bugpatterns.BugChecker.MethodInvocationTreeMatcher;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.suppliers.Supplier;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MemberSelectTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.tools.javac.code.Type;
+import java.util.Optional;
+import reactor.core.publisher.Mono;
+
+/**
+ * A {@link BugChecker} that flags usages of {@link Mono#zip(Mono, Mono)} and {@link
+ * Mono#zipWith(Mono)} with {@link Mono#empty()} parameters.
+ *
+ * <p>{@link Mono#zip(Mono, Mono)} and {@link Mono#zipWith(Mono)} perform incorrectly upon retrieval
+ * of the empty publisher and prematurely terminates the reactive chain from the execution. In most
+ * cases this is not the desired behaviour.
+ *
+ * @apiNote {@code Mono<?>.zipWith(Mono<Void>)} is allowed by the Reactor API, but it is an
+ *     incorrect usage of the API.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary =
+        "`Mono#zip` and `Mono#zipWith` should not be executed against `Mono#empty` or `Mono<Void>` parameter; "
+            + "please revisit the parameters used and make sure to supply correct publishers instead",
+    link = BUG_PATTERNS_BASE_URL + "MonoZipOfMonoVoidUsage",
+    linkType = CUSTOM,
+    severity = ERROR,
+    tags = LIKELY_ERROR)
+public final class MonoZipOfMonoVoidUsage extends BugChecker
+    implements MethodInvocationTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Supplier<Type> MONO_VOID_TYPE =
+      VisitorState.memoize(
+          generic(type("reactor.core.publisher.Mono"), type(Void.class.getCanonicalName())));
+  private static final Matcher<ExpressionTree> MONO_ZIP_AND_MONO_ZIP_WITH =
+      anyOf(
+          instanceMethod().onDescendantOf(MONO_VOID_TYPE).named("zipWith"),
+          staticMethod().onClass(MONO_VOID_TYPE).named("zip"));
+  private static final Matcher<ExpressionTree> GENERIC_ARGUMENT_DERIVED_FROM_MONO_TYPE =
+      toType(MethodInvocationTree.class, hasGenericArgumentOfExactType(MONO_VOID_TYPE));
+  private static final Matcher<ExpressionTree> HAS_MONO_EMPTY_AS_ARGUMENT =
+      toType(
+          MethodInvocationTree.class,
+          hasArguments(AT_LEAST_ONE, staticMethod().onClass(MONO_VOID_TYPE).named("empty")));
+  private static final Matcher<ExpressionTree> OPERATORS_WITH_MONO_VOID_GENERIC_ARGUMENT =
```

Comment:
Not this matcher specific, but I found we can use something from the `MoreTypePredicates#isSubType`.